### PR TITLE
[espnet3-recipe] Add SPGISpeech ASR recipe

### PR DESCRIPTION
**Pull request by chenehk**
Port of egs2/spgispeech/asr1 to ESPnet3. SPGISpeech is 5,000 h of transcribed financial earnings calls (1,966,109 utterances).

The recipe follows the egs3 layout used by librispeech_100 and mini_an4: run.py, conf/ (training, inference, metrics, publication), dataset/ (builder + Dataset), and src/ for the inference output formatter.

Corpus access is via the SPGISPEECH environment variable; nothing is downloaded automatically. dev_4k and train_nodev are derived by sorting train.csv by utterance id and splitting at 4,000 rows, mirroring egs2/spgispeech/asr1/local/data.sh.

Deviations from the egs2 recipe are marked [DEVIATION] in the training config and summarised in readme.md:

- Normalized transcripts rather than the `_unnorm` sets, because the dataset builder restricts the cache to val/dev_4k/train_nodev/train.
- No language model; ESPnet3 has no LM stage, while egs2 shallow-fuses one at lm_weight 0.3. Reported WER is therefore LM-free.
- batch_bins is not egs2's 35000000: egs2 bins on raw samples as a global budget, ESPnet3 bins on feats_shape (frames x 80) per GPU. The conversion is derived in the config.
- num_workers 4 rather than espnet2's default of 1.

Partial result on dev_4k (4,000 utts), valid.acc.ave_10best, beam 20, ctc_weight 0.3, no LM, after 25 of 35 epochs on 4 GPUs:

    CER 1.01    WER 2.42

No pretrained model is published with this PR. The evaluation is incomplete: val (39,341 utterances), which egs2 reports, has not been decoded, and training stopped short of the configured 35 epochs. A model will be packaged and uploaded once training completes and the full test sets are scored.

Claude-Session: https://claude.ai/code/session_01SUG8bJYjyQqUquCUk69rRE

## What did you change?

---

## Why did you make this change?

---

## Is your PR small enough?

---

### CODE REVIEW

File: egs3/spgispeech/asr/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+#
+# Self-contained: no `defaults:` block. The only base is
+# egs3/TEMPLATE/asr/conf/training.yaml, which run.py injects automatically via
+# load_and_merge_config(default_package="egs3.TEMPLATE.asr").
+#
+# Reference training environment:
+# - Number of GPUs: 4
+# - GPU type: NVIDIA GH200 (96 GB)
+# - Training time: approximately 25 h for 25 epochs
+# - dev_4k result at 25 epochs: WER 2.42 / CER 1.01 (beam 20, no LM)
```
Comment (sw005320) [inline egs3/spgispeech/asr/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml]: Is this correct?

File: egs3/spgispeech/asr/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+##########################################################
+# Keyword arguments forwarded verbatim to Lightning's Trainer.fit()
+# (espnet3/systems/base/training.py: `trainer.fit(**fit_kwargs)`); the model is
+# supplied by ESPnet3 itself. Empty means a fresh run.
+#
+# The useful key is `ckpt_path`, which resumes instead of starting over:
+#   fit:
+#     ckpt_path: last                  # ${exp_dir}/last.ckpt
+#     ckpt_path: /path/to/some.ckpt    # or an explicit checkpoint
+fit: {}
```
Comment (sw005320) [inline egs3/spgispeech/asr/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml]: I believe we do not need to.
@Masao-Someki, please confirm

**Comment by sw005320** [comment]
Please use batch decoding and confirm that the result will not change.

@Masao-Someki, please make batch decoding the default.

**Comment by sw005320** [comment]
See https://github.com/espnet/espnet/pull/6658
Once it is fixed, I'll ask you to incorporate batch decoding and rerun your evalaution

File: egs3/spgispeech/asr/dataset/builder.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+def iter_source_candidates(
+    recipe_root: Path,
+    source_dir: str | Path | None,
+) -> Iterable[Path]:
+    """Yield candidate directories that may be the SPGISpeech root."""
+    yield recipe_root / _CFG["dataset_path"] / "spgispeech"
+
+    if source_dir is not None:
+        yield Path(source_dir)
```
[the remark this answers is not in the document]
Comment (chenehk) [reply egs3/spgispeech/asr/dataset/builder.py]: Fixed in [5be734a](https://github.com/espnet/espnet/pull/6656/commits/5be734af5de4e0ab5fdbd7d554a7cf8fe3c97048)

File: egs3/spgispeech/asr/dataset/builder.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+            _call_supported(
+                self._build_recipe_source,
+                recipe_dir=recipe_dir,
+                cache=no_cache,
+                **kwargs,
+            )
+        _build_hf_cache(recipe_dir, cache_root, kwargs)
+
+
+_HF_CACHE_SPLITS = ["val", "dev_4k", "train_nodev", "train"]
```
[the remark this answers is not in the document]
Comment (chenehk) [reply egs3/spgispeech/asr/dataset/builder.py]: Fixed in [5be734a](https://github.com/espnet/espnet/pull/6656/commits/5be734af5de4e0ab5fdbd7d554a7cf8fe3c97048)

File: egs3/spgispeech/asr/conf/publication.yaml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  exclude:
+    # inference outputs (TEMPLATE defaults)
+    - inference
+    - inference_*
+    # every raw Lightning checkpoint: periodic_step*, step*, last*, epoch*.
+    # Inference loads the averaged .pth, never these.
+    - "*.ckpt"
+    # keep ONLY valid.acc.ave_10best.pth. "?" is a single character, so this
+    # matches ave_1best..ave_9best but not ave_10best.
+    - "valid.acc.ave_?best.pth"
```
[the remark this answers is not in the document]
Comment (chenehk) [reply egs3/spgispeech/asr/conf/publication.yaml:56]: Fixed in [af68d0e](https://github.com/espnet/espnet/pull/6656/commits/af68d0ed369e3b3b39c939130765e78cde98fa62)

File: egs3/spgispeech/asr/dataset/builder.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+        shutil.rmtree(temporary, ignore_errors=True)
+        failures = cache_root / f"{split}.failures.jsonl"
+
+        dataset = _call_supported(
+            dataset_class,
+            split=split,
+            recipe_dir=recipe_dir,
+            cache={"enabled": False},
+            **dataset_kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (chenehk) [reply egs3/spgispeech/asr/dataset/builder.py:339]: This was a temporary fix for a limited experiment. I've removed `max_utts` entirely in [2ca5f35](https://github.com/espnet/espnet/pull/6656/commits/2ca5f35b9288ed6e4fcfb2a76d83b78af17c8e8f)

File: egs3/spgispeech/asr/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml (hunk 2 above)
Comment (chenehk) [reply egs3/spgispeech/asr/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml]: I've removed it in [5be734a](https://github.com/espnet/espnet/pull/6656/commits/5be734af5de4e0ab5fdbd7d554a7cf8fe3c97048)

File: egs3/spgispeech/asr/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml (hunk 1 above)
Comment (chenehk) [reply egs3/spgispeech/asr/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml]: The details about GPU and training time was a little off, I fixed them in [af68d0e](https://github.com/espnet/espnet/pull/6656/commits/af68d0ed369e3b3b39c939130765e78cde98fa62). The dev_4k result is obtained by inference, not the training-time validation, so these are correct numbers

**Review by sw005320 — APPROVED** [review]

**Comment by sw005320** [comment]
I think this recipe looks good, and I merged the PR.
If you have additional experiments, please report them via the follow-up PR, including the batch decoding result.

**Comment by sw005320** [comment]
Follow-up on the batch-decoding check requested above, now that #6658 (batched decoding by default in espnet3), #6633 and #6660 (encoder outputs independent of batch padding) are in master.

I ran the check on this machine with the model-zoo SPGISpeech conformer (`espnet/Shinji_Watanabe_spgispeech_asr_train_asr_conformer6_n_fft512_hop_lengt-truncated-f1ac86`), which has the same architecture as this recipe's `train_asr_conformer6_n_fft512_hop_length256.yaml`, on 100 utterances of the ungated `macabdul9/spgispeech_S_eval_1K` set, beam 20, ctc_weight 0.3 (references lower-cased and stripped of punctuation for scoring, so the absolute WER is not comparable to the paper's).

| batch_size | WER | transcripts identical to batch_size 1 |
|---:|---:|---:|
| 1 | 8.05% | -- |
| 4 | 8.17% | 87 / 100 |
| 8 | 8.09% | 79 / 100 |
| 16 | 7.96% | 79 / 100 |

(utterances sorted by length; unsorted, batch 4 gave 7.34% and 79/100.)

**Batched decoding does not reproduce `batch_size 1` exactly for this model, and it cannot be made to.** The recipe follows egs2 in leaving `rel_pos_type` unset, so `ConformerEncoder` falls back to the legacy relative positional encoding (the config already notes the deprecation warning). In that implementation the position table is indexed by `(T - 1) + j - i` with `T` the *padded* length of the batch, so every frame of any utterance that is not the longest in its batch sees different positional encodings. I measured the encoder directly: with #6633 the output lengths no longer change in a batch (0 of 100), the longest utterance of each batch is bit-identical to decoding it alone (max |diff| 6e-6), and every padded utterance differs on every frame (median max |diff| 0.35). Nothing else is left: the subsampling-length and time-convolution effects were removed by #6633 and #6660, and with `rel_pos_type: latest` the same conformer is invariant to 1e-6.

The WER differences above are noise (a one-word change in 13-21 of 100 utterances, in both directions), so batched decoding is safe to use with this model; the transcripts just will not match to the utterance.

Two suggestions for the follow-up PR:

1. **For any new training, set `encoder_conf.rel_pos_type: latest`.** It changes the attention module and therefore the model, so it is not a drop-in for the existing checkpoint, but a model trained with it decodes identically at any batch size (the LibriSpeech E-Branchformer, which uses it, is 16/16 identical). This is the deviation from egs2 worth taking.
2. For the existing model, report the batch-decoded WER alongside `batch_size 1`, and note in `readme.md` that exact reproduction needs `batch_size: null`.

Unrelated to correctness but found while measuring: end detection was re-summarising every ended hypothesis at every step, which dominated `post_process` at larger batches; fixed in #6667.
